substrate-node/pallets/pallet-smart-contract/src/billing.rs

```diff
@@ -32,173 +38,300 @@ impl<T: Config> Pallet<T> {
+                let node =
+                    pallet_tfgrid::Nodes::<T>::get(contract.get_node_id()).ok_or_else(|| {
+                        log::error!("Node not exists for contract_id: {:?}", contract_id);
+                        Error::<T>::NodeNotExists
+                    })?;
+                let farm = pallet_tfgrid::Farms::<T>::get(node.farm_id).ok_or_else(|| {
+                    log::error!("Farm not exists for node_id: {:?}", node.farm_id);
+                    Error::<T>::FarmNotExists
+                })?;
+                let farmer_twin =
+                    pallet_tfgrid::Twins::<T>::get(farm.twin_id).ok_or_else(|| {
+                        log::error!("Twin not exists for farm_id: {:?}", farm.twin_id);
+                        Error::<T>::TwinNotExists
+                    })?;
+                (Some(farmer_twin), Some(node.certification))
+            } else {
+                (None, None)
+            };
 
-        let twin =
-            pallet_tfgrid::Twins::<T>::get(contract.twin_id).ok_or(Error::<T>::TwinNotExists)?;
-        let usable_balance = Self::get_usable_balance(&twin.account_id);
-        let stash_balance = Self::get_stash_balance(twin.id);
-        let total_balance = usable_balance
-            .checked_add(&stash_balance)
-            .unwrap_or(BalanceOf::<T>::zero());
+        // Switch lazily from old contract lock to use new contract payment state
+        // This allows tracking the overdrafted amount for the current contract
+        // While still avoiding the need for a storage migration for all contracts
+        let mut contract_payment_state = ContractPaymentState::<T>::get(contract.contract_id);
+        // Get and remove the contract lock from storage
+        let old_contract_lock = ContractLock::<T>::take(contract.contract_id);
+        // This is no-op if the contract lock is empty (migrated)
+        Self::ensure_contract_migrated(
+            &src_twin.account_id,
+            &old_contract_lock,
+            &mut contract_payment_state,
+        );
 
-        let now = Self::get_current_timestamp_in_secs();
+        // Calculate user total usable balance
+        let twin_usable_balance = Self::get_usable_balance(&src_twin.account_id);
+        let stash_usable_balance = Self::get_stash_balance(src_twin.id);
+        let total_usable_balance =
+            twin_usable_balance.defensive_saturating_add(stash_usable_balance);
+        let now: u64 = Self::get_current_timestamp_in_secs();
 
-        // Calculate amount of seconds elapsed based on the contract lock struct
-        let mut contract_lock = ContractLock::<T>::get(contract.contract_id);
-        let seconds_elapsed = now.checked_sub(contract_lock.lock_updated).unwrap_or(0);
+        // Calculate amount of seconds elapsed based on the contract payment state
+        let seconds_elapsed =
+            now.defensive_saturating_sub(contract_payment_state.last_updated_seconds);
```

Comment:
Both using `checked_sub` with `unwrap_or(0)` or `saturating_sub` will yield the same result.
However, `saturating_sub` makes more sense here for its simplicity since it inherently saturates preventing underflow without the need for additional error-handling logic like `unwrap_or(0)`.

Then, in Substrate, we have the `defensive_` variants like `defensive_saturating_sub` which will log a warning message when underflow or overflow occurs to help you detect such bugs in production but otherwise behave as usual. Also, it will panic if accrued in tests (which is an added benefit).

So, instead of having something like that
```
let seconds_elapsed = now.checked_sub(contract_lock.lock_updated).unwrap_or_else(|| {
        // Log a warning about the underflow.
        warn!("Underflow detected: now ({}) is less than last_updated ({}). Returning 0.", now, contract_lock.lock_updated);
        // Return 0 in case of underflow.
        0
    })
 ```

we can replace it with just
```
let seconds_elapsed =
            now.defensive_saturating_sub(contract_payment_state.last_updated_seconds);
```